File: c++/src/silo.cpp
#include "TurningGrilleCracker.hpp"

#include "BlockingPortionQueue.hpp"
#include "NonBlockingQueue.hpp"
#include <voidland/blown_queue/BlownQueue.hpp>

#include <fstream>
#include <string>
#include <iostream>
#include <list>
#include <thread>
#include <chrono>
#include <atomic>
#include <memory>
#include <boost/algorithm/string.hpp>
#include <boost/core/demangle.hpp>


using namespace voidland::concurrent;


inline static const std::string CIPHER_TEXT_PATH = "encrypted_msg.txt";
inline static const std::string CLEAR_TEXT_PATH = "decrypted_msg.txt";


static void heatCpu()
{
    unsigned cpuCount = std::thread::hardware_concurrency();
    std::vector<std::thread> workerThreads;
    workerThreads.reserve(cpuCount);

    std::atomic<bool> stop(false);

    for (unsigned i = 0; i < cpuCount; i++)
    {
        workerThreads.push_back(std::thread
            {
                [&stop]
                {
                    while (!stop.load(std::memory_order_relaxed));
                }
            });
    }

    std::this_thread::sleep_for(std::chrono::minutes(1));
    stop.store(true, std::memory_order_relaxed);

    for (std::thread& workerThread : workerThreads)
    {
        workerThread.join();
    }
}

int main(int argc, char *argv[])
{
    try
    {
        std::string cipherText;
        {
            std::ifstream cipherTextStream(CIPHER_TEXT_PATH);
            if (cipherTextStream.fail())
            {
                throw std::ios::failure("Cannot open file: " + CIPHER_TEXT_PATH);
            }

            std::getline(cipherTextStream, cipherText);
        }
        std::string clearText;
        {
            std::ifstream clearTextStream(CLEAR_TEXT_PATH);
            if (clearTextStream.fail())
            {
                throw std::ios::failure("Cannot open file: " + CLEAR_TEXT_PATH);
            }

            std::getline(clearTextStream, clearText);
        }
        boost::to_upper(clearText);
        boost::erase_all_regex(clearText, turning_grille::NOT_CAPITAL_ENGLISH_LETTERS_RE);

        std::string arg;
        if (argc >= 2)
        {
            arg = argv[1];
            boost::to_lower(arg);
        }
        else
        {
            arg = "syncless";
        }

        std::unique_ptr<turning_grille::TurningGrilleCrackerImplDetails> crackerImplDetails;
        
        unsigned cpuCount = std::thread::hardware_concurrency();
        if (arg == "concurrent")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		queue::createConcurrentBlownQueue<turning_grille::Grille>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "atomic")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		queue::createAtomicBlownQueue<turning_grille::Grille>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "lockfree")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		queue::createLockfreeBlownQueue<turning_grille::Grille>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "onetbb")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::NonBlockingQueue<turning_grille::Grille>> nonBlockingQueue =
            		std::make_unique<queue::OneTBB_PortionQueue<turning_grille::Grille>>(maxQueueSize);
            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::BlownQueue<turning_grille::Grille>>(maxQueueSize, std::move(nonBlockingQueue));
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "michael_scott")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		queue::createMichaelScottBlownQueue<turning_grille::Grille>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "ramalhete")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		queue::createRamalheteBlownQueue<turning_grille::Grille>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "vyukov")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		queue::createVyukovBlownQueue<turning_grille::Grille>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "kirsch_1fifo")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::NonBlockingQueue<turning_grille::Grille>> nonBlockingQueue =
            		std::make_unique<queue::Kirsch1FifoPortionQueue<turning_grille::Grille>>(maxQueueSize);
            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::BlownQueue<turning_grille::Grille>>(maxQueueSize, std::move(nonBlockingQueue));
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "kirsch_bounded_1fifo")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::NonBlockingQueue<turning_grille::Grille>> nonBlockingQueue =
            		std::make_unique<queue::KirschBounded1FifoPortionQueue<turning_grille::Grille>>(maxQueueSize);
            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::BlownQueue<turning_grille::Grille>>(maxQueueSize, std::move(nonBlockingQueue));
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "nikolaev")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::NonBlockingQueue<turning_grille::Grille>> nonBlockingQueue =
            		std::make_unique<queue::NikolaevPortionQueue<turning_grille::Grille>>(maxQueueSize);
            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::BlownQueue<turning_grille::Grille>>(maxQueueSize, std::move(nonBlockingQueue));
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "nikolaev_bounded")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::NonBlockingQueue<turning_grille::Grille>> nonBlockingQueue =
            		std::make_unique<queue::NikolaevBoundedPortionQueue<turning_grille::Grille>>(maxQueueSize);
            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::BlownQueue<turning_grille::Grille>>(maxQueueSize, std::move(nonBlockingQueue));
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "onetbb_bounded")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::OneTBB_BoundedPortionQueue<turning_grille::Grille>>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "sync_bounded")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::SyncBoundedPortionQueue<turning_grille::Grille>>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "textbook")
        {
            unsigned initialConsumerCount = cpuCount * 3;
            unsigned producerCount = cpuCount;
            std::size_t maxQueueSize = initialConsumerCount * producerCount * 1000;

            std::unique_ptr<queue::MPMC_PortionQueue<turning_grille::Grille>> portionQueue =
            		std::make_unique<queue::TextbookPortionQueue<turning_grille::Grille>>(maxQueueSize);
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerProducerConsumer>(initialConsumerCount, producerCount, std::move(portionQueue));
        }
        else if (arg == "syncless")
        {
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerSyncless>();
        }
        else if (arg == "serial")
        {
            crackerImplDetails = std::make_unique<turning_grille::TurningGrilleCrackerSerial>();
        }
        else
        {
            throw std::invalid_argument("Unexpected argument: " + arg);
        }

        turning_grille::TurningGrilleCracker cracker(cipherText, std::move(crackerImplDetails));
        if (!VERBOSE)
        {
        	heatCpu();
        }
        std::set<std::string> candidates = cracker.bruteForce();

        if (candidates.find(clearText) == candidates.end())
        {
        	throw std::logic_error("The correct clear text was not found among the decrypted candidates.");
        }
        return 0;
    }
    catch (const std::exception& e)
    {
        std::cerr << boost::core::demangle(typeid(e).name()) << ": " << e.what() << std::endl;
        return 1;
    }
}

File: rust/src/silo.rs
#![allow(non_snake_case)]
#![allow(non_camel_case_types)]

#![feature(mpmc_channel)]

pub mod org;

use org::voidland::concurrent::VERBOSE;
use org::voidland::concurrent::queue::MostlyNonBlockingPortionQueue;
use org::voidland::concurrent::queue::MPMC_PortionQueue;
use org::voidland::concurrent::queue::StdTextbookPortionQueue;
use org::voidland::concurrent::queue::ParkingLotTextbookPortionQueue;
use org::voidland::concurrent::queue::SyncMpmcPortionQueue;
use org::voidland::concurrent::turning_grille::NOT_CAPITAL_ENGLISH_LETTERS_RE;
use org::voidland::concurrent::turning_grille::Grille;
use org::voidland::concurrent::turning_grille::TurningGrilleCracker;
use org::voidland::concurrent::turning_grille::TurningGrilleCrackerImplDetails;
use org::voidland::concurrent::turning_grille::TurningGrilleCrackerProducerConsumer;
use org::voidland::concurrent::turning_grille::TurningGrilleCrackerSyncless;
use org::voidland::concurrent::turning_grille::TurningGrilleCrackerSerial;

use std::fs::read_to_string;
use std::string::String;
use std::env;
use std::thread::available_parallelism;
use std::sync::atomic::AtomicBool;
use std::sync::atomic::Ordering;
use std::sync::Arc;
use std::thread;
use std::thread::JoinHandle;
use std::time;
use std::collections::BTreeSet;


const CIPHER_TEXT_PATH: &str = "encrypted_msg.txt";
const CLEAR_TEXT_PATH: &str = "decrypted_msg.txt";


fn heatCpu()
{
    let cpuCount: usize = available_parallelism().unwrap().get();
    let mut workerThreads: Vec<JoinHandle<_>> = Vec::new();
    
    let stop: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));
    
    for _i in 0..cpuCount
    {
        let stop = stop.clone();
        workerThreads.push(thread::spawn(move ||
            {
                while !stop.load(Ordering::Relaxed)
                {
                }
            }));
    }
    
    thread::sleep(time::Duration::from_secs(60));
    stop.store(true, Ordering::Relaxed);
    
    for workerThread in &mut workerThreads.into_iter()
    {
        let _ = workerThread.join().unwrap();
    }
}

fn main()
{
    let cipherText: String = read_to_string(CIPHER_TEXT_PATH).unwrap().lines().next().unwrap().to_string();
    let clearText: String = read_to_string(CLEAR_TEXT_PATH).unwrap().lines().next().unwrap().to_string();
    let clearText: String = clearText.to_uppercase();
    let clearText: String = NOT_CAPITAL_ENGLISH_LETTERS_RE.replace_all(&clearText, "").to_string();
    
    let args: Vec<String> = env::args().collect();
    let arg: &str;
    if args.len() >= 2
    {
        arg = &args[1];
    }
    else
    {
        arg = "syncless";
    }
    
    let crackerImplDetails: Box<dyn TurningGrilleCrackerImplDetails>;

    let cpuCount: usize = available_parallelism().unwrap().get();
    match arg
    {
        "concurrent" =>
        {
            let initialConsumerCount: usize = cpuCount * 3;
            let producerCount: usize = cpuCount;
            let maxQueueSize: usize = initialConsumerCount * producerCount * 1000;

            let portionQueue: Arc<dyn MPMC_PortionQueue<Grille>> = Arc::new(MostlyNonBlockingPortionQueue::createConcurrentBlownQueue(maxQueueSize));
            crackerImplDetails = Box::new(TurningGrilleCrackerProducerConsumer::new(initialConsumerCount, producerCount, portionQueue));
        }
        "async_mpmc" =>
        {
            let initialConsumerCount: usize = cpuCount * 3;
            let producerCount: usize = cpuCount;
            let maxQueueSize: usize = initialConsumerCount * producerCount * 1000;

            let portionQueue: Arc<dyn MPMC_PortionQueue<Grille>> = Arc::new(MostlyNonBlockingPortionQueue::createAsynMPMC_BlownQueue(maxQueueSize));
            crackerImplDetails = Box::new(TurningGrilleCrackerProducerConsumer::new(initialConsumerCount, producerCount, portionQueue));
        }
        "textbook" =>
        {
            let initialConsumerCount: usize = cpuCount * 3;
            let producerCount: usize = cpuCount;
            let maxQueueSize: usize = initialConsumerCount * producerCount * 1000;

            let portionQueue: Arc<dyn MPMC_PortionQueue<Grille>> = Arc::new(StdTextbookPortionQueue::new(maxQueueSize));
            crackerImplDetails = Box::new(TurningGrilleCrackerProducerConsumer::new(initialConsumerCount, producerCount, portionQueue));
        }
        "textbook_pl" =>
        {
            let initialConsumerCount: usize = cpuCount * 3;
            let producerCount: usize = cpuCount;
            let maxQueueSize: usize = initialConsumerCount * producerCount * 1000;

            let portionQueue: Arc<dyn MPMC_PortionQueue<Grille>> = Arc::new(ParkingLotTextbookPortionQueue::new(maxQueueSize));
            crackerImplDetails = Box::new(TurningGrilleCrackerProducerConsumer::new(initialConsumerCount, producerCount, portionQueue));
        }
        "sync_mpmc" =>
        {
            let initialConsumerCount: usize = cpuCount * 3;
            let producerCount: usize = cpuCount;
            let maxQueueSize: usize = initialConsumerCount * producerCount * 1000;

            let portionQueue: Arc<dyn MPMC_PortionQueue<Grille>> = Arc::new(SyncMpmcPortionQueue::new(maxQueueSize));
            crackerImplDetails = Box::new(TurningGrilleCrackerProducerConsumer::new(initialConsumerCount, producerCount, portionQueue));
        }
        "syncless" =>
        {
            crackerImplDetails = Box::new(TurningGrilleCrackerSyncless::new());
        }
        "serial" =>
        {
            crackerImplDetails = Box::new(TurningGrilleCrackerSerial::new());
        }
        _ =>
        {
            panic!("Unexpected argument: {}", arg);
        }
    }
    
    let cracker: Arc<TurningGrilleCracker> = Arc::new(TurningGrilleCracker::new(&cipherText, crackerImplDetails));
    if !*VERBOSE
    {
        heatCpu();
    }
    let candidates: BTreeSet<String> = cracker.bruteForce();
    
    if !candidates.contains(&clearText)
    {
        panic!("The correct clear text was not found among the decrypted candidates.");
    }
}
